File: frontend/types/video.ts
export interface SubtitleTrack {
  lang: string;
  name: string;
  ext: string;
}

export interface VideoMetadata {
  id: string;
  url: string;
  title: string;
  channel: string;
  channel_url?: string;
  duration: number;
  duration_string: string;
  thumbnail: string;
  description?: string;
  upload_date?: string;
  view_count?: number;
  max_resolution: number;
  available_resolutions: number[];
  available_video_formats: string[];
  available_audio_formats: string[];
  subtitles: SubtitleTrack[];
  is_live: boolean;
}

export interface AnalyzeResponse {
  metadata: VideoMetadata;
  already_downloaded: boolean;
  existing_record?: DownloadRecord;
}

export interface DownloadRecord {
  id: string;
  video_id: string;
  url: string;
  title: string;
  channel: string;
  thumbnail: string;
  format: string;
  quality: string;
  file_path: string;
  file_size?: number;
  status: string;
  created_at: string;
  completed_at?: string;
  error_message?: string;
}

File: backend/app/models/video.py
from typing import List, Optional
from pydantic import BaseModel, Field


class SubtitleTrack(BaseModel):
    lang: str
    name: str
    ext: str = "vtt"


class VideoMetadata(BaseModel):
    id: str
    url: str
    title: str
    channel: str
    channel_url: Optional[str] = None
    duration: int = 0
    duration_string: str = "00:00"
    thumbnail: str = ""
    description: Optional[str] = None
    upload_date: Optional[str] = None
    view_count: Optional[int] = None
    max_resolution: int = 1080
    available_resolutions: List[int] = Field(default_factory=list)
    available_video_formats: List[str] = Field(default_factory=lambda: ["mp4", "webm", "mkv"])
    available_audio_formats: List[str] = Field(default_factory=lambda: ["mp3", "m4a", "opus", "wav"])
    subtitles: List[SubtitleTrack] = Field(default_factory=list)
    is_live: bool = False
